Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+{
+	Release();
+}
+
+void CUIItemUpgrade::Release()
+{
+	for (int i = 0; i < MAX_ITEM_UPGRADE_SLOT; i++)
+	{
+		DeleteIconItemSkill(m_pMyUpgradeSLot[i]);
+		m_iUpgradeSlotInvPos[i+1] = -1; // 0 is UpgradeItemSlot [1,2..9] m_pMyUpgradeSLot slot
+	}
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		DeleteIconItemSkill(m_pMyUpgradeInv[i]);
+		DeleteIconItemSkill(m_pBackupUpgradeInv[i]);
+	}
+	DeleteIconItemSkill(m_pUpgradeItemSlot);
+	m_iUpgradeSlotInvPos[0] = -1; //UpgradeItemSlot position
+
+	if (m_pUITooltipDlg)
+	{
+		delete m_pUITooltipDlg;
+		m_pUITooltipDlg = nullptr;
+	}
+	m_pStrMyGold = nullptr;
+
+	CN3UIBase::Release();
+}
+
+void CUIItemUpgrade::Tick()
+{
+	switch (m_eAnimationState)
+	{
+		case ANIM_START:
+			StartUpgradeAnim();
+			break;
+		case ANIM_FLIPFLOP:
+			UpdateFlipFlopAnimation();
+			break;
+		case ANIM_RESULT:
+			ShowResultUpgrade();
+			break;
+		case ANIM_COVER_OPENING:
+			UpdateCoverAnimation();
+			break;
+		case ANIM_NONE:
+			m_fAnimationTimer = 0.0f;
+			m_iCurrentFrame = 0;
+			break;
+		default:
+			break;
+	}
+
+	CN3UIBase::Tick();
+}
+
+void CUIItemUpgrade::Render()
+{
+	if (!m_bVisible) 
+		return;
+
+	POINT ptCur = CGameProcedure::s_pLocalInput->MouseGetPos();
+
+	if (m_pUITooltipDlg)
+		m_pUITooltipDlg->DisplayTooltipsDisable();
+
+	bool bTooltipRender = false;
+	__IconItemSkill* spItem = nullptr;
+
+	for (UIListReverseItor itor = m_Children.rbegin(); m_Children.rend() != itor; ++itor)
+	{
+		CN3UIBase* pChild = (*itor);
+		if ((GetState() == UI_STATE_ICON_MOVING) && (pChild->UIType() == UI_TYPE_ICON) && (CN3UIWndBase::m_sSelectedIconInfo.pItemSelect) &&
+			((CN3UIIcon*) pChild == CN3UIWndBase::m_sSelectedIconInfo.pItemSelect->pUIIcon))
+			continue;
+		pChild->Render();
+
+		if ((GetState() == UI_STATE_COMMON_NONE) &&
+			(pChild->UIType() == UI_TYPE_ICON) && (pChild->GetStyle() & UISTYLE_ICON_HIGHLIGHT))
+		{
+			bTooltipRender = true;
+			spItem = GetHighlightIconItem((CN3UIIcon*) pChild);
+		}
+	}
+
+	// Display the count for items that should show a count.
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (m_pMyUpgradeInv[i] && ((m_pMyUpgradeInv[i]->pItemBasic->byContable == UIITEM_TYPE_COUNTABLE) ||
+			(m_pMyUpgradeInv[i]->pItemBasic->byContable == UIITEM_TYPE_COUNTABLE_SMALL)))
+		{
+			char szID[32];
+			sprintf(szID, "s_count_%d", i);
+			CN3UIString* pStr = (CN3UIString*)GetChildByID(szID);
+			if (pStr)
+			{
+				if ((GetState() == UI_STATE_ICON_MOVING) && (m_pMyUpgradeInv[i] == CN3UIWndBase::m_sSelectedIconInfo.pItemSelect))
+				{
+					pStr->SetVisible(false);
+				}
+				else
+				{
+					if (m_pMyUpgradeInv[i]->pUIIcon->IsVisible())
+					{
+						pStr->SetVisible(true);
+						pStr->SetStringAsInt(m_pMyUpgradeInv[i]->iCount);
+						pStr->Render();
+					}
+					else
+					{
+						pStr->SetVisible(false);
+					}
+				}
+			}
+		}
+		else
+		{
+			char szID[32];
+			sprintf(szID, "s_count_%d", i);
+			CN3UIString* pStr = (CN3UIString*)GetChildByID(szID);
```

Comment:
You're performing a scan across every child in this UI to find the control for every single string control for each item, every single frame. This is a performance killer.
Pre-load these in `Load()` (ideally) or `InitIconWnd()` (typically used by `CN3UIWndBase` classes). `Load()` is preferable.